File: haskell/p21.hs
-- P21 Insert an element at a given position into a list.

-- Using "splitAt"
f1 :: Int -> a -> [a] -> [a]
f1 n x ys = let (a, b) = splitAt (pred n) ys in a ++ x : b

-- By index (inefficient)
f2 :: Int -> a -> [a] -> [a]
f2 n x ys = take (pred n) ys ++ x : drop (pred n) ys

-- Recursion
f3 :: Int -> a -> [a] -> [a]
f3 1 x ys = x : ys
f3 n x (y:ys) = y : f3 (pred n) x ys

-- Tail recursion
f4 :: Int -> a -> [a] -> [a]
f4 = f []
    where f acc 1 x ys = reverse acc ++ x : ys
          f acc n x (y:ys) = f (y:acc) (pred n) x ys

File: clojure/P21.clj
; P21 Insert an element at a given position into a list.

; Using "split-at"
(defn f1 [n x l]
  (let [[a b] (split-at (dec n) l)]
    (concat a (cons x b))))

; By index (inefficient)
(defn f2 [n x l]
  (concat (take (dec n) l) (cons x (drop (dec n) l))))

; Recursion
(defn f3 [n x l]
  (if (= n 1)
    (cons x l)
    (cons (first l) (f3 (dec n) x (rest l)))))

; Tail recursion
(defn f4 [n x l]
  (loop [i n ys l acc '()]
    (if (= i 1)
      (concat (reverse acc) (cons x ys))
      (recur (dec i) (rest ys) (cons (first ys) acc)))))